synapse/storage/util/id_generators.py

```diff
@@ -352,12 +356,39 @@ def _mark_id_as_finished(self, next_id: int):
 
         with self._lock:
             self._unfinished_ids.discard(next_id)
+            self._finished_ids.add(next_id)
+
+            new_cur = None
+
+            if self._unfinished_ids:
+                # If there are unfinished IDs then the new position will be the
+                # largest finished ID less than the minimum unfinished ID.
```

Comment:
I'm unsure if it is clearer, but I think something like:

```python
new_cur = min_finished
for s in self._finished_ids:
    if s < min_funished:
        new_cur = max(new_cur, s)
    else:
        finished.add(s)
```

But that might make the handling later on harder.

Reply:
Potentially, but then you iterate over the list twice (since you're doing a `min(...)`)?